src/entry.c

```diff
@@ -0,0 +1,385 @@
+             *     +------+-------+---------------+
+             *     | TTL  | value | field         |
+             *     |      | ptr   | hdr8 "foo" \0 |
+             *     +------+-------+---------------+
+             */
+            embed_value = false;
+            alloc_size += sizeof(sds);
+            if (embedded_field_sds_type == SDS_TYPE_5) {
+                embedded_field_sds_type = SDS_TYPE_8;
+                alloc_size -= field_alloc_size;
+                field_alloc_size = sdsReqSize(field_len, embedded_field_sds_type);
+                alloc_size += field_alloc_size;
+            }
+        }
+    }
+    if (expiry_size) *expiry_size = expiry_alloc_size;
+    if (field_sds_type) *field_sds_type = embedded_field_sds_type;
+    if (field_size) *field_size = field_alloc_size;
+    if (embedded_value_size) *embedded_value_size = embedded_value_alloc_size;
+    if (is_value_embedded) *is_value_embedded = embed_value;
+
+    return alloc_size;
+}
+
+/* Takes ownership of value. does not take ownership of field */
+entry *entryCreate(const_sds field, sds value, long long expiry) {
+    bool embed_value = false;
+    int embedded_field_sds_type;
+    size_t expiry_size, embedded_value_sds_size, embedded_field_sds_size;
+    size_t alloc_size = entryReqSize(field, value, expiry, &embed_value, &embedded_field_sds_type, &embedded_field_sds_size, &expiry_size, &embedded_value_sds_size);
+    size_t buf_size;
+
+    /* allocate the buffer */
+    char *buf = zmalloc_usable(alloc_size, &buf_size);
+
+    /* Set The expiry if exists */
+    if (expiry_size) {
+        *(long long *)buf = expiry;
+        buf += expiry_size;
+        buf_size -= expiry_size;
+    }
+    if (value) {
+        if (!embed_value) {
+            *(sds *)buf = value;
+            buf += sizeof(sds);
+            buf_size -= sizeof(sds);
+        } else {
+            sdswrite(buf + embedded_field_sds_size, buf_size - embedded_field_sds_size, SDS_TYPE_8, value, sdslen(value));
+            sdsfree(value);
+            buf_size -= embedded_value_sds_size;
+        }
+    }
+    /* Set the field data */
+    entry *new_entry = sdswrite(buf, embedded_field_sds_size, embedded_field_sds_type, field, sdslen(field));
+
+    /* Field sds aux bits are zero, which we use for this entry encoding. */
+    sdsSetAuxBit(new_entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR, embed_value ? 0 : 1);
+    sdsSetAuxBit(new_entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_EXPIRY, expiry_size > 0 ? 1 : 0);
+
+    /* Check that the new entry was built correctly */
+    debugServerAssert(sdsGetAuxBit(new_entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR) == (embed_value ? 0 : 1));
+    debugServerAssert(sdsGetAuxBit(new_entry, FIELD_SDS_AUX_BIT_ENTRY_HAS_EXPIRY) == (expiry_size > 0 ? 1 : 0));
+    return new_entry;
+}
+
+/* Modify the entry's value and/or expiration time.
+ * In case the provided value is NULL, will use the existing value. */
+entry *entryUpdate(entry *e, sds value, long long expiry) {
```

Comment:
I'd still recommend passing in a pointer to `value` to indicate that owership can be transferred (`value` nullified on return); otherwise at the minimum, we should make it clear in the comment that `value`'s ownership can be transferred. 

```suggestion
entry *entryUpdate(entry *e, sds *value, long long expiry) {
```

Reply:
I am njot against this concept. I think it can help prevent some programing mistakes. I would just like to say that it does have impact on the CURRENT flows. the entry module is simply a copy of the existing hashTypeEntry which I wanted to generalize. 
in the existing code the entry API was taking the ownership:
```
/* takes ownership of value, does not take ownership of field */
hashTypeEntry *hashTypeCreateEntry(sds field, sds value) {
```

and this allows calling code to simplify a bit by having a one-liners like:
`hashTypeEntry *e = hashTypeCreateEntry(field, sdsdup(value));`

So I would prefer to avoid making more external code changes at this point. however if you feel this will add much value I can make the effort.

Reply:
Mention ownership in the comment then.

Reply:
done @zuiderkwast